### PR TITLE
[OTel] Introduce otel subcommand

### PR DESCRIPTION
**Pull request by michalpristas**
This PR introduces `otel` subcommand to elastic agent.

Pre PR agent detected mode based on a name passed using a config flag. If it was otel/otelcol/otlp it switched to otel mode otherwise it used normal agent mode. This functionality is no longer present.

When running using `run` command, pre-otel behavior is used regardless of config file name.

When using `otel` subcommand agent starts in otel mode keeping `status` subcommand  working. 
In this PR no other subcommand are added such as validate. 
multiple `--config` flags are allowed with values supported with normal collector as described here: [how-to-provide-configuration](https://github.com/open-telemetry/opentelemetry-collector/tree/main/service#how-to-provide-configuration)

`--set` is supported for overriding config options as described here [--set](https://github.com/open-telemetry/opentelemetry-collector/tree/main/service#the-format-and-limitations-of---set)

Fixes: https://github.com/elastic/elastic-agent/issues/3903
Fixes: https://github.com/elastic/elastic-agent/issues/4080
Fixes: https://github.com/elastic/elastic-agent/issues/3870

### CODE REVIEW

**Comment by michalpristas** [comment]
`status` and `run` are covered by existing integ tests

**Comment by ycombinator** [comment]
I tested this PR with a few different ways of specifying OTel configuration and they all worked as expected.  Adding a few "Fixes: #xxxx" to the issue description to reflect this.

File: internal/pkg/agent/cmd/otel_flags.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	var sets []string
+	setVals, err := cmd.Flags().GetStringArray(setFlagName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve set flags: %w", err)
+	}
+
+	for _, s := range setVals {
+		idx := strings.Index(s, "=")
+		if idx == -1 {
+			// No need for more context, see TestSetFlag/invalid_set.
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/otel_flags.go:45]: > see TestSetFlag/invalid_set

Where is this test case?

File: internal/pkg/agent/cmd/otel_flags.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+			return nil, fmt.Errorf("missing equal sign for set value %q", s)
+		}
+		sets = append(sets, "yaml:"+strings.TrimSpace(strings.ReplaceAll(s[:idx], ".", "::"))+": "+strings.TrimSpace(s[idx+1:]))
+	}
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/otel_flags.go:49]: Nice to have: extract into own function(s) and add unit tests.

File: internal/pkg/agent/cmd/otel.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			})
+		},
+		SilenceUsage:  true,
+		SilenceErrors: true,
+	}
+
+	cmd.SetHelpFunc(func(c *cobra.Command, s []string) {
+		c.InheritedFlags().VisitAll(func(f *pflag.Flag) {
+			f.Hidden = true
+		})
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/otel.go:50]: Nit: repeated in `PreRun` above; consider extracting into common function, assuming the reason for this hiding behavior is the same in both places.

Comment (ycombinator) [reply internal/pkg/agent/cmd/otel.go:50]: Also, would be good to add a comment as to why we're hiding inherited flags.

File: internal/pkg/agent/cmd/run.go [hunk 4]
```diff
@@ -208,7 +177,7 @@ func runElasticAgent(ctx context.Context, cancel context.CancelFunc, override cf
 	}
 
 	// agent ID needs to stay empty in bootstrap mode
-	createAgentID := true
+	createAgentID := !runAsOtel
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/run.go:180]: Just double checking: this change is not strictly related to this PR but a fix for a small bug you found while working on this PR?

File: internal/pkg/agent/cmd/run.go [hunk 5]
```diff
@@ -405,7 +374,11 @@ LOOP:
 
 func loadConfig(ctx context.Context, override cfgOverrider, runAsOtel bool) (*configuration.Configuration, error) {
 	if runAsOtel {
-		return configuration.DefaultConfiguration(), nil
+		defaultCfg := configuration.DefaultConfiguration()
+		// disable monitoring to avoid injection
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/run.go:378]: Appreciate the comment here but I didn't follow what you mean by "avoid injection".  Could you expand on this a bit in the comment, please?

File: internal/pkg/otel/run.go [hunk 6]
```diff
@@ -95,3 +74,11 @@ func newSettings(configPath string, version string) (*otelcol.CollectorSettings,
 		DisableGracefulShutdown: true,
 	}, nil
 }
+
+func makeMapProvidersMap(providers ...confmap.Provider) map[string]confmap.Provider {
```
Comment (ycombinator) [inline internal/pkg/otel/run.go:78]: Easy candidate for a unit test?

File: internal/pkg/agent/application/paths/common.go [hunk 7]
```diff
@@ -20,6 +20,8 @@ import (
 const (
 	// DefaultConfigName is the default name of the configuration file.
 	DefaultConfigName = "elastic-agent.yml"
+	// DefaultConfigName is the default name of the configuration file.
```
Comment (ycombinator) [inline internal/pkg/agent/application/paths/common.go:23]: Comment needs to be updated.

File: internal/pkg/agent/application/paths/common.go [hunk 8]
```diff
@@ -146,8 +148,19 @@ func SetConfig(path string) {
 
 // ConfigFile returns the path to the configuration file.
 func ConfigFile() string {
+
+	return configFileWithDefaultOverride(DefaultConfigName)
+}
+
+// ConfigFile returns the path to the configuration file.
```
Comment (ycombinator) [inline internal/pkg/agent/application/paths/common.go:155]: Comment needs to be updated.

File: internal/pkg/agent/application/paths/common.go [hunk 9]
```diff
@@ -146,8 +148,19 @@ func SetConfig(path string) {
...
+func OtelConfigFile() string {
+	return configFileWithDefaultOverride(DefaultOtelConfigName)
+}
+
+// ConfigFile returns the path to the configuration file.
```
Comment (ycombinator) [inline internal/pkg/agent/application/paths/common.go:160]: Comment needs to be updated.

**Comment by ycombinator** [comment]
This is looking good.  Left a few minor comments/questions.

File: internal/pkg/agent/cmd/run.go (hunk 4 above)
Comment (michalpristas) [reply internal/pkg/agent/cmd/run.go:180]: yes, on mac it fails to run otel because agent tries to create agent id and do not have permissions if not running as root

File: internal/pkg/agent/cmd/run.go (hunk 5 above)
Comment (michalpristas) [reply internal/pkg/agent/cmd/run.go:378]: updated a comment. it's not really an issue now as we're passing `inputs: []`
added this just to be sure and have default reflecting use case a bit more

**Review by ycombinator — APPROVED** [review]
LGTM.

**Comment by michalpristas** [comment]
@cmacknz this is similar case to PR force-merged recently.
do you see some requirements missing preventing proceeding with the same path?

**Review by cmacknz — APPROVED** [review]
The integration tests use the `otel` sub-command so I'm fine to bypass the unit test coverage check.
